src/toil_scripts/gatk_germline/germline.py

```diff
@@ -1,304 +1,500 @@
-                                                  genome_id, mock=mock).rv()
+    genome_id = config.genome_fasta
+    if not hasattr(config, 'genome_fai') or config.genome_fai is None:
+        config.genome_fai = job.addChildJobFn(run_samtools_faidx,
+                                              genome_id,
+                                              cores=config.cores).rv()
+    if not hasattr(config, 'genome_dict') or config.genome_dict is None:
+        config.genome_dict = job.addChildJobFn(run_picard_create_sequence_dictionary,
+                                               genome_id,
+                                               cores=config.cores).rv()
     return config
 
 
-def gatk_haplotype_caller(job, bam_id, bai_id, config, annotations=None):
+def prepare_bam(job, uuid, url, config, rg_line=None):
+    """
+    Prepares BAM file for GATK Germline pipeline.
+
+    0: Align FASTQ or Download BAM
+    1: Remove secondary alignments
+    2: Sort
+    3: Index
+
+    :param JobFunctionWrappingJob job: Toil Job instance
+    :param str uuid: Unique identifier for the sample
+    :param str url: URL to BAM file
+    :param Namespace config: Configuration options for pipeline
+    :param str rg_line: RG line for BWA alignment. Default is None.
+    :return: BAM and BAI FileStoreIDs
+    :rtype: tuple
+    """
+    # Determine extension on input file
+    base, ext1 = os.path.splitext(url)
+    _, ext2 = os.path.splitext(base)
+
+    # Produce or download BAM file
+    if {ext1, ext2} & {'.fq', '.fastq'}:
+        if not config.run_bwa:
+            raise ValueError('Not configured to run BWA. Please set run-bwa: True')
+        elif not rg_line:
+            raise ValueError('FASTQ file requires an RG line for BWA alignment')
+        else:
+            get_bam = job.wrapJobFn(setup_and_run_bwa_kit, url, config).encapsulate()
+
+    elif {ext1, ext2} & {'.bam'}:
+        job.fileStore.logToMaster("Downloading BAM: %s" % uuid)
+        get_bam = job.wrapJobFn(download_url_job,
+                                url,
+                                name='toil.bam',
+                                synapse_login=config.synapse_login,
+                                s3_key_path=config.ssec,
+                                disk=config.file_size).encapsulate()
+    else:
+        raise ValueError('Accepted formats: FASTQ or BAM\n'
+                         'Could not determine {} file type from URL:\n{}'.format(uuid, url))
+
+    rm_secondary = job.wrapJobFn(run_samtools_view,
+                                 get_bam.rv(),
+                                 flag='0x800',
+                                 ncores=config.cores,
+                                 disk=config.file_size, cores=config.cores)
+
+    sort_bam_disk = PromisedRequirement(lambda x: 3 * x.size + human2bytes('10G'),
+                                        rm_secondary.rv())
+    sort_bam = job.wrapJobFn(run_samtools_sort,
```

Comment:
You may want to make sorting optional, since someone may provide a sorted BAM as input.